File: AcePump.WebApi/Startup/LogRequestAttribute.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Web.Http.Controllers;
using System.Web.Http.Filters;
using AcePump.Common;
using AcePump.Domain;
using AcePump.Domain.Models;

namespace AcePump.WebApi.Startup
{
    public class LogRequestAttribute : ActionFilterAttribute
    {
        public override void OnActionExecuting(HttpActionContext actionContext)
        {
            if (AcePumpEnvironment.Environment.Configuration.Logging.LogAllRequests)
            {
                using (var db = DataSourceFactory.GetAcePumpDataSource())
                {
                    var request = new Log_HttpRequest {
                        Environment = "webapi",
                        HttpMethod = actionContext.Request.Method.Method,
                        Path = actionContext.Request.RequestUri.AbsolutePath,
                        RequestTime = DateTime.Now,
                        LoggedInUsername = actionContext.RequestContext.Principal.Identity.IsAuthenticated ? actionContext.RequestContext.Principal.Identity.Name : "",
                        Parameters = new List<Log_HttpRequestParam>()
                    };

                    foreach (KeyValuePair<string, string> qstringEntry in actionContext.Request.GetQueryNameValuePairs())
                    {
                        if (qstringEntry.Key.Equals("password", StringComparison.InvariantCultureIgnoreCase)) continue;

                        request.Parameters.Add(new Log_HttpRequestParam
                        {
                            ParamType = "qstring",
                            ParamName = qstringEntry.Key,
                            ParamValue = qstringEntry.Value
                        });
                    }
                    
                    foreach (KeyValuePair<string, object> actionArgument in actionContext.ActionArguments)
                    {
                        if (actionArgument.Key.Equals("password", StringComparison.InvariantCultureIgnoreCase)) continue;

                        request.Parameters.Add(new Log_HttpRequestParam
                        {
                            ParamType = "actionargument",
                            ParamName = actionArgument.Key,
                            ParamValue = actionArgument.Value.ToString()
                        });
                    }

                    foreach (KeyValuePair<string, object> routeDataEntry in actionContext.RequestContext.RouteData.Values)
                    {
                        request.Parameters.Add(new Log_HttpRequestParam {
                            ParamType = "routedata",
                            ParamName = routeDataEntry.Key,
                            ParamValue = routeDataEntry.Value.ToString()
                        });
                    }

                    db.Log_HttpRequests.Add(request);
                    db.SaveChanges();
                }
            }
        }
    }
}

File: AcePump.Web/ActionFilters/LogRequestAttribute.vb
Imports AcePump.Common
Imports AcePump.Domain
Imports AcePump.Domain.DataSource
Imports AcePump.Domain.Models

Namespace ActionFilters
    Public Class LogRequestAttribute
        Inherits ActionFilterAttribute

        Public Overrides Sub OnActionExecuting(filterContext As ActionExecutingContext)
            If AcePumpEnvironment.Environment.Configuration.Logging.LogAllRequests Then
                Using db As AcePumpContext = DataSourceFactory.GetAcePumpDataSource()
                    Dim request = New Log_HttpRequest With {
                        .Environment = "mvc",
                        .HttpMethod = filterContext.HttpContext.Request.HttpMethod,
                        .Path = filterContext.HttpContext.Request.Path,
                        .RequestTime = DateTime.Now,
                        .LoggedInUsername = If(filterContext.HttpContext.User.Identity.IsAuthenticated, filterContext.HttpContext.User.Identity.Name, ""),
                        .Parameters = New List(Of Log_HttpRequestParam)
                    }

                    For Each qstringKey As String In filterContext.HttpContext.Request.QueryString
                        If qstringKey.Equals("password", StringComparison.InvariantCultureIgnoreCase) Then Continue For

                        For Each qstringValue As String In filterContext.HttpContext.Request.QueryString.GetValues(qstringKey)
                            request.Parameters.Add(New Log_HttpRequestParam With {
                                .ParamType = "qstring",
                                .ParamName = qstringKey,
                                .ParamValue = qstringValue
                            })
                        Next
                    Next

                    For Each formKey As String In filterContext.HttpContext.Request.Form
                        If formKey.Equals("password", StringComparison.InvariantCultureIgnoreCase) Then Continue For

                        For Each formValue As String In filterContext.HttpContext.Request.Form.GetValues(formKey)
                            request.Parameters.Add(New Log_HttpRequestParam With {
                                .ParamType = "form",
                                .ParamName = formKey,
                                .ParamValue = formValue
                            })
                        Next
                    Next

                    For Each routeDataEntry As KeyValuePair(Of String, Object) In filterContext.RequestContext.RouteData.Values
                        request.Parameters.Add(New Log_HttpRequestParam With {
                            .ParamType = "routedata",
                            .ParamName = routeDataEntry.Key,
                            .ParamValue = routeDataEntry.Value
                        })
                    Next

                    db.Log_HttpRequests.Add(request)
                    db.SaveChanges()
                End Using
            End If
        End Sub
    End Class
End Namespace